### PR TITLE
Added time filter in the Notes View

### PR DESCRIPTION
**Pull request by altair1673**
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

I have added a DateTimeRange Picker Accordion component to the Notes View. This component is in JavaFX unlike the one in the Data Access View which is a Swing component. The user can specify a datetime range to see a specific set of notes. The User can change time zones. They have to toggle the on/off button of the component and then hit apply for their filters to work. If the user clicks apply with the "active: button set to off then all notes are displayed.
### Why Should This Be In Core?

Users can sort notes created at specific times. Allows for less clutter in the Notes View if only the important can be filtered.
The component is in JavaFX so some elements can be put in separate classes for future use whenever another time range picker needs to be used.
### Benefits

Allows more control to the user for which notes they want to see.
### Possible Drawbacks

The placement of the component might need to change, currently it is at the top of  the Notes View and when clicked it expands down making the top toolbar even bigger.
~~The dropdown for the time zones is extremely large and there is no scroll bar so it takes a while to scroll to the bottom. In the future maybe this can be replaced with a search box. Options also don't have the time offset next to them, eg. no "GMT+10" next to the region names which could be an issue.~~
~~The comparison code of the datetime variable of a note entry is dependent on the date format used in the Notes View. If the date formatting were to change then the code to filter by datetime would break and throw exceptions. This needs to change in the future~~
### Verification Process

The component successfully filters notes even if the time zone is different than the one the notes were created in.
### Applicable Issues

#1761

### CODE REVIEW

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]
Having both an on/off button and an apply button is very confusing. Only one should be needed here:
image
If there are too many notes present in the Notes View then half of the time range selector is not accessible and gets cut off
image

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    public Pane getPane() {
+        return dateTimePane;
+    }
+
+    /**
+     * Sets the current date time to whatever the current local time is
+     *
+     * @param zone - zone ID of current location
+     */
+    public void setCurrentDateTime(ZoneId zone) {
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:110]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        minPicker.getValueFactory().setValue(timeAtZone.getMinute());
+        secPicker.getValueFactory().setValue(timeAtZone.getSecond());
+    }
+
+    /**
+     * Converts time to another time zone
+     *
+     * @param convertTo - id of zone to convert to
+     */
+    public void convertCurrentDateTime(ZoneId convertTo) {
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:126]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 3]
```diff
@@ -32,41 +32,62 @@
...
 import javafx.scene.input.KeyCode;
 import javafx.scene.input.KeyEvent;
+import javafx.scene.layout.Background;
+import javafx.scene.layout.BackgroundFill;
 import javafx.scene.layout.BorderPane;
+import javafx.scene.layout.GridPane;
 import javafx.scene.layout.HBox;
+import javafx.scene.layout.Pane;
 import javafx.scene.layout.Priority;
+import javafx.scene.layout.TilePane;
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:87]: Lots of unused imports

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+ * limitations under the License.
+ */
+package au.gov.asd.tac.constellation.views.notes;
+
+import java.time.LocalDate;
+import java.time.Month;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import javafx.beans.property.IntegerProperty;
+import javafx.beans.property.SimpleIntegerProperty;
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:23]: Unused imports

**Review by Delphinus8821 — CHANGES_REQUESTED** [review]

**Comment by Delphinus8821** [comment]
It would be good to have some unit tests be written for the DateTimePicker class

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import javafx.scene.control.TitledPane;
+import javafx.scene.layout.GridPane;
+import javafx.scene.layout.Pane;
+import javafx.scene.paint.Color;
+
+/**
+ *
+ * @author altair1673
+ */
+public class DateTimeRangePicker {
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java:39]: Need to add a class description here

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        clearButton.setOnMouseEntered(event -> clearButton.setStyle("-fx-background-color: #23FFB5; "));
+
+        clearButton.setOnMouseExited(event -> clearButton.setStyle("-fx-background-color: #7FFFD4;  "));
+
+        applyButton.setOnMouseEntered(event -> applyButton.setStyle("-fx-background-color: #078BC9; "));
+
+        applyButton.setOnMouseExited(event -> applyButton.setStyle("-fx-background-color: #0080FF; "));
+
+
+    }
```
Comment (Delphinus8821) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java:158]: This spacing can be minimised

**Review by Delphinus8821 — APPROVED** [review]

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    private ZoneId zone;
+
+    public DateTimePicker(final boolean from) {
+        dateTimePane = new Pane();
+        this.from = from;
+        dateTimePane.getChildren().add(mainGridPane);
+
+        GridPane datePickerGridPane = new GridPane();
+
+        Label datePickerLabel = new Label();
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:56]: final x2

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+
+        datePickerGridPane.add(datePickerLabel, 0, 0);
+        datePickerGridPane.add(datePicker, 1, 0);
+
+        GridPane timePickerGrid = new GridPane();
+        Label hourLabel = new Label("Hour");
+        Label minLabel = new Label("Minute");
+        Label secLabel = new Label("Second");
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:71]: final x4

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+    /**
+     * Sets the current date time to whatever the current local time is
+     *
+     * @param zone - zone ID of current location
+     */
+    public void setCurrentDateTime(final ZoneId zone) {
+        this.zone = zone;
+
+        ZonedDateTime timeAtZone = ZonedDateTime.now(zone);
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimePicker.java:110]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+     *
+     * @param entryTime - Time that is being checked
+     * @return whether entryTime is within range
+     */
+    public boolean checkIsWithinRange(ZonedDateTime entryTime) {
+        // Convert entry time to time zone of DateTimePickers
+        entryTime = entryTime.withZoneSameInstant(fromDate.getZoneId());
+
+        ZonedDateTime fromTime = fromDate.getCurrentDateTime();
+        ZonedDateTime toTime = toDate.getCurrentDateTime();
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/DateTimeRangePicker.java:216]: final x2

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 11]
```diff
@@ -492,16 +519,52 @@ protected synchronized void updateNotesUI() {
...
+                    // Get date time of entry in proper format
+                    String dateFormat = new SimpleDateFormat(DATETIME_PATTERN).format(new Date(Long.parseLong(entry.getDateTime())));
+
+                    // Extract date components
+                    String[] dateTimeComponents = dateFormat.split(" ");
+                    String time = dateTimeComponents[0];
+                    String date = dateTimeComponents[3];
+
+                    // Split time into hour, minute and day
+                    String[] timeComponents = time.split(":");
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:532]: final x5

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 12]
```diff
@@ -492,16 +519,52 @@ protected synchronized void updateNotesUI() {
...
+                    int hour = Integer.parseInt(timeComponents[0]);
+                    int min = Integer.parseInt(timeComponents[1]);
+                    int sec = Integer.parseInt(timeComponents[2]);
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:535]: final x2

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 13]
```diff
@@ -492,16 +519,52 @@ protected synchronized void updateNotesUI() {
...
+                    }
+
+                    // Split date into day, month and year
+                    String[] dateComponents = date.split("/");
+                    int day = Integer.parseInt(dateComponents[0]);
+                    int month = Integer.parseInt(dateComponents[1]);
+                    int year = Integer.parseInt(dateComponents[2]);
+
+                    // Convert time of notes to user specified time zone
+                    ZonedDateTime entryTime = ZonedDateTime.of(year, month, day, hour, min, sec, 0, ZoneId.of(ZoneId.systemDefault().getId()));
```
Comment (OrionsGuardian) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:550]: final

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Test of setCurrentDateTime method, of class DateTimePicker.
+     */
+    @Test
+    public void testSetCurrentDateTime() {
+        System.out.println("setCurrentDateTime");
+        ZoneId zone = ZoneId.of("Australia/Perth");
+        ZonedDateTime currentTime = ZonedDateTime.now(zone);
+        DateTimePicker instance = new DateTimePicker(true);
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:76]: final x3

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Test of convertCurrentDateTime method, of class DateTimePicker.
+     */
+    @Test
+    public void testConvertCurrentDateTime() {
+        System.out.println("convertCurrentDateTime");
+        ZoneId convertTo = ZoneId.of("Australia/Perth");
+        ZoneId currentZone = ZoneId.of("Australia/Perth");
+        DateTimePicker instance = new DateTimePicker(true);
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:92]: final x3

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Test of isActive method, of class DateTimePicker.
+     */
+    @Test
+    public void testIsActive() {
+        System.out.println("isActive");
+        DateTimePicker instance = new DateTimePicker(true);
+        boolean expResult = false;
+        boolean result = instance.isActive();
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:113]: final x3

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Test of setActive method, of class DateTimePicker.
+     */
+    @Test
+    public void testSetActive() {
+        System.out.println("setActive");
+        boolean active = true;
+        DateTimePicker instance = new DateTimePicker(true);
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:126]: final x2

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+        assertEquals(instance.isActive(), true);
+    }
+
+    /**
+     * Test of getCurrentDateTime method, of class DateTimePicker.
+     */
+    @Test
+    public void testGetCurrentDateTime() {
+        System.out.println("getCurrentDateTime");
+        DateTimePicker instance = new DateTimePicker(true);
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:138]: final

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+        instance.setCurrentDateTime(ZoneId.of("Australia/Adelaide"));
+        ZonedDateTime expResult = ZonedDateTime.now(ZoneId.of("Australia/Adelaide"));
+        ZonedDateTime result = instance.getCurrentDateTime();
+
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:142]: final x2

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /**
+     * Test of getZoneId method, of class DateTimePicker.
+     */
+    @Test
+    public void testGetZoneId() {
+        System.out.println("getZoneId");
+        DateTimePicker instance = new DateTimePicker(true);
+        ZoneId zone = ZoneId.of("Australia/Perth");
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:155]: final x2

File: CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+        instance.setCurrentDateTime(zone);
+
+        ZoneId result = instance.getZoneId();
```
Comment (OrionsGuardian) [inline CoreNotesView/test/unit/src/au/gov/asd/tac/constellation/views/notes/DateTimePickerNGTest.java:158]: final

**Review by OrionsGuardian — CHANGES_REQUESTED** [review]

**Comment by OrionsGuardian** [comment]
The new functionality seems fine, however, I would like the layout of the Date Picker component to be modified a litttle:

image

From the above mock sample, I'd like the width of the Date Picker to be reduced, and the Clear button to be moved to titlebar, and only available when a filter is active.

**Review by OrionsGuardian — APPROVED** [review]

**Pull request merged**